File: scripts/qa/metrics.py
"""Agreement metrics for accuracy-audit analysis (stdlib only).

1:1 mirror (names + semantics) of the Rust library
`crates/lbm-core/tests/common/metrics.rs`. Rust is the source of truth for
in-suite assertions; this module serves post-hoc analysis of CLI/scenario
output (compose with the parsers in qa_checks.py). Keep the two in sync —
if you change one, change the other in the same commit.

Run self-tests: python3 scripts/qa/metrics.py
"""

import math
from typing import Callable, List, NamedTuple, Sequence, Tuple


def l2_rel(actual: Sequence[float], reference: Sequence[float]) -> float:
    """Relative L2 error ||a - r||_2 / ||r||_2."""
    assert len(actual) == len(reference)
    num = sum((a - r) ** 2 for a, r in zip(actual, reference))
    den = sum(r * r for r in reference)
    return math.sqrt(num / den)


def linf_rel(actual: Sequence[float], reference: Sequence[float],
             floor: float) -> float:
    """Relative L-infinity error: max|a-r| / max(max|r|, floor)."""
    assert len(actual) == len(reference)
    dmax = max(abs(a - r) for a, r in zip(actual, reference))
    rmax = max(max(abs(r) for r in reference), floor)
    return dmax / rmax


class LinFit(NamedTuple):
    """y ~ slope*x + intercept; r2 rejects sloppy fits (assert it too)."""
    slope: float
    intercept: float
    r2: float


def linear_fit(x: Sequence[float], y: Sequence[float]) -> LinFit:
    assert len(x) == len(y) and len(x) >= 2
    n = float(len(x))
    mx, my = sum(x) / n, sum(y) / n
    sxx = sum((xi - mx) ** 2 for xi in x)
    sxy = sum((xi - mx) * (yi - my) for xi, yi in zip(x, y))
    syy = sum((yi - my) ** 2 for yi in y)
    slope = sxy / sxx
    intercept = my - slope * mx
    r2 = 1.0 if syy == 0.0 else (sxy * sxy) / (sxx * syy)
    return LinFit(slope, intercept, r2)


def order_fit(h: Sequence[float], err: Sequence[float]) -> LinFit:
    """Observed convergence order (log-log slope). Assert slope AND r2."""
    return linear_fit([math.log(v) for v in h], [math.log(v) for v in err])


def envelope_fit(y: Sequence[float], amp: Sequence[float]) -> LinFit:
    """amp ~ A*exp(-k*y): fit of ln(amp) on y; A = exp(intercept), k = -slope."""
    return linear_fit(list(y), [math.log(v) for v in amp])


def phase_fit(t: Sequence[float], signal: Sequence[float],
              omega: float) -> Tuple[float, float]:
    """(amplitude, phase) with signal ~ amplitude*sin(omega*t + phase).

    Quadrature projection; sample an integer number of periods.
    """
    assert len(t) == len(signal) and t
    n = float(len(t))
    s = sum(si * math.sin(omega * ti) for ti, si in zip(t, signal))
    c = sum(si * math.cos(omega * ti) for ti, si in zip(t, signal))
    a_sin, a_cos = 2.0 * s / n, 2.0 * c / n
    return math.hypot(a_sin, a_cos), math.atan2(a_cos, a_sin)


def monotonicity(xs: Sequence[float]) -> float:
    """Fraction of adjacent pairs strictly decreasing (1.0 = monotone decay)."""
    assert len(xs) >= 2
    dec = sum(1 for a, b in zip(xs, xs[1:]) if b < a)
    return dec / (len(xs) - 1)


class CurveAgreement(NamedTuple):
    max_rel_dev: float
    worst_x: float
    frac_in_band: float


def curve_agreement(theory: Callable[[float], float],
                    samples: Sequence[Tuple[float, float]],
                    rel_band: float, floor: float) -> CurveAgreement:
    """A3 primitive: each sample must lie ON the theory curve, point by point."""
    assert samples
    max_rel_dev, worst_x, in_band = 0.0, samples[0][0], 0
    for x, measured in samples:
        th = theory(x)
        dev = abs(measured - th) / max(abs(th), floor)
        if dev > max_rel_dev:
            max_rel_dev, worst_x = dev, x
        if dev <= rel_band:
            in_band += 1
    return CurveAgreement(max_rel_dev, worst_x, in_band / len(samples))


def _selftest() -> None:
    # Mirrors crates/lbm-core/tests/metrics_selftest.rs fixtures.
    assert abs(l2_rel([1, 2.3, 3, 4], [1, 2, 3, 4]) - 0.3 / math.sqrt(30)) < 1e-15
    assert abs(linf_rel([1, 2.3, 3, 4], [1, 2, 3, 4], 0.0) - 0.3 / 4) < 1e-15
    assert abs(linf_rel([0.1], [0.0], 1.0) - 0.1) < 1e-15

    h: List[float] = [0.1, 0.05, 0.025, 0.0125]
    fit = order_fit(h, [3 * x * x for x in h])
    assert abs(fit.slope - 2.0) < 1e-12 and fit.r2 > 1 - 1e-12
    assert order_fit(h, [3e-2, 8e-3, 5e-3, 4.8e-3]).r2 < 0.98

    y = [0.5, 1.0, 2.0, 3.5, 5.0]
    fit = envelope_fit(y, [0.7 * math.exp(-0.35 * v) for v in y])
    assert abs(fit.slope + 0.35) < 1e-12
    assert abs(math.exp(fit.intercept) - 0.7) < 1e-12

    omega = 2 * math.pi / 100
    t = list(range(400))
    amp, phase = phase_fit(t, [0.02 * math.sin(omega * ti + 0.6) for ti in t], omega)
    assert abs(amp - 0.02) < 1e-6 and abs(phase - 0.6) < 1e-6

    assert monotonicity([4, 3, 2, 1]) == 1.0
    assert abs(monotonicity([4, 3, 3.5, 1]) - 2 / 3) < 1e-15

    agree = curve_agreement(lambda x: x * x,
                            [(1, 1), (2, 4), (3, 9.9), (4, 16)], 0.05, 0.0)
    assert abs(agree.max_rel_dev - 0.1) < 1e-12
    assert agree.worst_x == 3 and abs(agree.frac_in_band - 0.75) < 1e-12
    print("metrics.py selftest OK")


if __name__ == "__main__":
    _selftest()

File: crates/lbm-core/tests/common/metrics.rs
//! Agreement metrics for accuracy-audit and validation tests.
//!
//! Pure functions only: no simulation types, no I/O. Every accuracy assertion
//! in an adversarial test should go through one of these so the semantics are
//! reviewable in one place. Mirrored 1:1 (names + semantics) by
//! `scripts/qa/metrics.py` for post-hoc analysis of CLI output.

/// Relative L2 norm of the error between `actual` and `reference`:
/// `||a - r||_2 / ||r||_2`.
pub fn l2_rel(actual: &[f64], reference: &[f64]) -> f64 {
    assert_eq!(actual.len(), reference.len());
    let mut num = 0.0;
    let mut den = 0.0;
    for (a, r) in actual.iter().zip(reference) {
        num += (a - r) * (a - r);
        den += r * r;
    }
    (num / den).sqrt()
}

/// Relative L-infinity error: `max |a - r| / max(max |r|, floor)`.
/// `floor` guards near-zero reference fields (pass 0.0 to disable).
pub fn linf_rel(actual: &[f64], reference: &[f64], floor: f64) -> f64 {
    assert_eq!(actual.len(), reference.len());
    let mut dmax = 0.0f64;
    let mut rmax = floor;
    for (a, r) in actual.iter().zip(reference) {
        dmax = dmax.max((a - r).abs());
        rmax = rmax.max(r.abs());
    }
    dmax / rmax
}

/// Least-squares straight-line fit `y ≈ slope·x + intercept`, with the
/// coefficient of determination `r2` so callers can reject sloppy fits
/// (an order estimate from a bad fit is meaningless — assert `r2` too).
#[derive(Clone, Copy, Debug)]
pub struct LinFit {
    pub slope: f64,
    pub intercept: f64,
    pub r2: f64,
}

pub fn linear_fit(x: &[f64], y: &[f64]) -> LinFit {
    assert_eq!(x.len(), y.len());
    assert!(x.len() >= 2, "linear_fit needs >= 2 points");
    let n = x.len() as f64;
    let mx = x.iter().sum::<f64>() / n;
    let my = y.iter().sum::<f64>() / n;
    let mut sxx = 0.0;
    let mut sxy = 0.0;
    let mut syy = 0.0;
    for (&xi, &yi) in x.iter().zip(y) {
        sxx += (xi - mx) * (xi - mx);
        sxy += (xi - mx) * (yi - my);
        syy += (yi - my) * (yi - my);
    }
    let slope = sxy / sxx;
    let intercept = my - slope * mx;
    let r2 = if syy == 0.0 { 1.0 } else { (sxy * sxy) / (sxx * syy) };
    LinFit {
        slope,
        intercept,
        r2,
    }
}

/// Observed convergence order from resolutions `h` and errors `err`
/// (log-log fit; slope = order). Assert BOTH `fit.slope` against the
/// theoretical order band AND `fit.r2` (>= ~0.98) — points off a straight
/// log-log line mean the asymptotic regime was not reached.
pub fn order_fit(h: &[f64], err: &[f64]) -> LinFit {
    let lx: Vec<f64> = h.iter().map(|v| v.ln()).collect();
    let ly: Vec<f64> = err.iter().map(|v| v.ln()).collect();
    linear_fit(&lx, &ly)
}

/// Exponential-envelope fit `amp ≈ A·exp(-k·y)` (semilog fit of ln(amp) on y).
/// Returns the fit on ln(amp): amplitude `A = intercept.exp()`, decay
/// `k = -slope`. Used for e.g. Stokes-II wall-layer decay and acoustic
/// damping envelopes.
pub fn envelope_fit(y: &[f64], amp: &[f64]) -> LinFit {
    let ly: Vec<f64> = amp.iter().map(|v| v.ln()).collect();
    linear_fit(y, &ly)
}

/// Single-frequency quadrature projection of `signal(t)` on `sin/cos(omega t)`:
/// returns `(amplitude, phase)` with `signal ≈ amplitude·sin(omega·t + phase)`.
/// Sampling should cover an integer number of periods for clean orthogonality.
pub fn phase_fit(t: &[f64], signal: &[f64], omega: f64) -> (f64, f64) {
    assert_eq!(t.len(), signal.len());
    assert!(!t.is_empty());
    let n = t.len() as f64;
    let mut s = 0.0;
    let mut c = 0.0;
    for (&ti, &si) in t.iter().zip(signal) {
        s += si * (omega * ti).sin();
        c += si * (omega * ti).cos();
    }
    let a_sin = 2.0 * s / n; // coefficient of sin(omega t)
    let a_cos = 2.0 * c / n; // coefficient of cos(omega t)
    ((a_sin * a_sin + a_cos * a_cos).sqrt(), a_cos.atan2(a_sin))
}

/// Fraction of adjacent pairs that are strictly decreasing (1.0 = the whole
/// sequence decreases monotonically). Use for error-vs-resolution and
/// error-vs-time sequences where theory demands monotone decay.
pub fn monotonicity(xs: &[f64]) -> f64 {
    assert!(xs.len() >= 2, "monotonicity needs >= 2 points");
    let dec = xs.windows(2).filter(|w| w[1] < w[0]).count();
    dec as f64 / (xs.len() - 1) as f64
}

/// Result of comparing measured samples against a theoretical curve.
#[derive(Clone, Copy, Debug)]
pub struct CurveAgreement {
    /// Largest relative deviation |measured - theory| / max(|theory|, floor).
    pub max_rel_dev: f64,
    /// x at which the largest deviation occurs.
    pub worst_x: f64,
    /// Fraction of samples within `rel_band` of the curve.
    pub frac_in_band: f64,
}

/// A3-axis primitive: the measured error/observable must lie ON the known
/// theoretical curve, point by point — "small" is not a pass. `floor` guards
/// zero crossings of the theory curve (pass 0.0 to disable).
pub fn curve_agreement(
    theory: impl Fn(f64) -> f64,
    samples: &[(f64, f64)],
    rel_band: f64,
    floor: f64,
) -> CurveAgreement {
    assert!(!samples.is_empty());
    let mut max_rel_dev = 0.0f64;
    let mut worst_x = samples[0].0;
    let mut in_band = 0usize;
    for &(x, measured) in samples {
        let th = theory(x);
        let dev = (measured - th).abs() / th.abs().max(floor);
        if dev > max_rel_dev {
            max_rel_dev = dev;
            worst_x = x;
        }
        if dev <= rel_band {
            in_band += 1;
        }
    }
    CurveAgreement {
        max_rel_dev,
        worst_x,
        frac_in_band: in_band as f64 / samples.len() as f64,
    }
}
